receiver/filereceiver/watcher.go

```diff
@@ -0,0 +1,135 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package filereceiver // import "github.com/open-telemetry/opentelemetry-collector-contrib/receiver/filereceiver"
+
+import (
+	"bufio"
+	"context"
+	"os"
+	"regexp"
+
+	"github.com/fsnotify/fsnotify"
+	"go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/model/otlp"
+	"go.uber.org/multierr"
+	"go.uber.org/zap"
+)
+
+var (
+	tracesUnmarshaler  = otlp.NewJSONTracesUnmarshaler()
+	logsUnmarshaler    = otlp.NewJSONLogsUnmarshaler()
+	metricsUnmarshaler = otlp.NewJSONMetricsUnmarshaler()
+)
+
+type watcher struct {
+	logger          *zap.Logger
+	path            string
+	watcher         *fsnotify.Watcher
+	logsConsumer    consumer.Logs
+	tracesConsumer  consumer.Traces
+	metricsConsumer consumer.Metrics
+}
+
+var (
+	logsFingerPrint    = regexp.MustCompile("\\{\\s*\"resourceLogs\"")
+	metricsFingerPrint = regexp.MustCompile("\\{\\s*\"resourceMetrics\"")
+	tracesFingerPrint  = regexp.MustCompile("\\{\\s*\"resourceSpans\"")
+)
+
+func (w *watcher) start() error {
```

Comment:
This should accept a `context.Context` so that the goroutine it spawns can be cancelled when the context is done.

Reply:
not sure the watcher start context should be tied in to the whole run, since the context seems to be the start context?

Reply:
That `context.Context` value ultimately comes from [`collector.Run()`](https://github.com/open-telemetry/opentelemetry-collector/blob/93272a5485166043651080ae9e89e75d9788d06b/service/collector.go#L227) and lives through the entire collector lifecycle.

Reply:
Ah, I had no idea. Why do we have a stop method then? Could context be a better way to dictate the behavior?